File: example/ios/UncompressExampleTests/Spy/ExtractableSpy.swift
//
//  ExtractableSpy.swift
//  UncompressExampleTests
//
//  Created by Adriano Souza Costa on 19/03/21.
//

import Foundation
import Uncompress

final class ExtractableSpy: Extractable {
  
  var compatibilities: [Compatibility] = []
  var filePath: URL?
  var destination: URL?
  var overwrite: Bool?
  var password: String?
  var isProtected: Bool?
  var error: Error?

  func extract(
    _ filePath: URL,
    to destination: URL,
    overwrite: Bool,
    password: String?
  ) throws {
    self.filePath = filePath
    self.destination = destination
    self.overwrite = overwrite
    self.password = password
    
    if let error = error {
      throw error
    }
  }
  
  func isProtected(_ filePath: URL) throws -> Bool {
    self.filePath = filePath
    
    if let error = error {
      throw error
    }
    
    return isProtected ?? false
  }

}

File: android/src/test/java/com/uncompress/spy/ExtractableSpy.kt
package com.uncompress.spy

import com.uncompress.data.intefaces.Extractable
import com.uncompress.domain.enum.Compatibility

class ExtractableSpy() : Extractable {
  override var compatibilities: List<Compatibility> = listOf()
  var filePath: String? = null
  var destination: String? = null
  var override: Boolean? = null
  var password: String? = null
  var isProtected: Boolean = false
  var error: Error? = null

  @Throws
  override fun extract(filePath: String, destination: String, override: Boolean, password: String?) {
    this.filePath = filePath
    this.destination = destination
    this.override = override
    this.password = password

    error?.let {
      throw it
    }
  }

  override fun isProtected(filePath: String): Boolean {
    this.filePath = filePath

    error?.let {
      throw it
    }

    return isProtected
  }
}
